src/tools/clippy/clippy_lints/src/collapsible_match.rs

```diff
@@ -49,104 +49,87 @@ declare_lint_pass!(CollapsibleMatch => [COLLAPSIBLE_MATCH]);
-        // one of the branches must be "wild-like"
-        if let Some(wild_inner_arm_idx) = arms_inner.iter().rposition(|arm_inner| is_wild_like(cx, &arm_inner.pat.kind, &arm_inner.guard));
-        let (wild_inner_arm, non_wild_inner_arm) =
-            (&arms_inner[wild_inner_arm_idx], &arms_inner[1 - wild_inner_arm_idx]);
-        if !pat_contains_or(non_wild_inner_arm.pat);
+        if let Some(binding_id) = path_to_local(peel_ref_operators(cx, inner_scrutinee));
+        if !pat_contains_or(inner_then_pat);
         // the binding must come from the pattern of the containing match arm
         // ..<local>.. => match <local> { .. }
         if let Some(binding_span) = find_pat_binding(outer_pat, binding_id);
-        // the "wild-like" branches must be equal
-        if wild_outer_block.map(|el| SpanlessEq::new(cx).eq_expr(wild_inner_arm.body, el)).unwrap_or(true);
+        // the "else" branches must be equal
+        if match (outer_else_body, inner_else_body) {
+            (None, None) => true,
+            (None, Some(e)) | (Some(e), None) => is_unit_expr(e),
+            (Some(a), Some(b)) => SpanlessEq::new(cx).eq_expr(a, b),
+        };
         // the binding must not be used in the if guard
         let mut used_visitor = LocalUsedVisitor::new(cx, binding_id);
-        if match outer_guard {
-            None => true,
-            Some(Guard::If(expr) | Guard::IfLet(_, expr)) => !used_visitor.check_expr(expr),
+        if outer_guard.map_or(true, |(Guard::If(e) | Guard::IfLet(_, e))| !used_visitor.check_expr(e));
+        // ...or anywhere in the inner expression
+        if match inner {
+            IfLetOrMatch::IfLet(_, _, body, els) => {
+                !used_visitor.check_expr(body) && els.map_or(true, |e| !used_visitor.check_expr(e))
+            },
+            IfLetOrMatch::Match(_, arms, ..) => !arms.iter().any(|arm| used_visitor.check_arm(arm)),
         };
-        // ...or anywhere in the inner match
-        if !arms_inner.iter().any(|arm| used_visitor.check_arm(arm));
         then {
-            span_lint_and_then(
-                cx,
-                COLLAPSIBLE_MATCH,
-                expr.span,
-                "unnecessary nested match",
-                |diag| {
-                    let mut help_span = MultiSpan::from_spans(vec![binding_span, non_wild_inner_arm.pat.span]);
-                    help_span.push_span_label(binding_span, "replace this binding".into());
-                    help_span.push_span_label(non_wild_inner_arm.pat.span, "with this pattern".into());
-                    diag.span_help(help_span, "the outer pattern can be modified to include the inner pattern");
-                },
+            let what = |is_match| if is_match { "match" } else { "if let" };
+            let msg = format!(
+                "this `{}` can be collapsed into the outer `{}`",
+                what(matches!(inner, IfLetOrMatch::Match(..))),
+                what(outer_else_body.is_some()),
```

Comment:
An outer if-let block can also have an else block, meaning that this is not a clear indication that the outer body is a `match` statement. This can lead to wrong lint messages:

##### Input `src/tools/clippy/tests/ui/collapsible_match.rs:32:9`
```rs
    // if let else, if let else
    if let Ok(val) = res_opt {
        if let Some(n) = val {
            take(n);
        } else {
            return;
        }
    } else {
        return;
    }
```

##### Output `src/tools/clippy/tests/ui/collapsible_match.stderr`
```txt
help: the outer pattern can be modified to include the inner pattern
  --> $DIR/collapsible_match.rs:24:15
   |
LL |     if let Ok(val) = res_opt {
   |               ^^^ replace this binding
LL |         if let Some(n) = val {
   |                ^^^^^^^ with this pattern

error: this `if let` can be collapsed into the outer `match`
  --> $DIR/collapsible_match.rs:32:9
   |
LL | /         if let Some(n) = val {
LL | |             take(n);
LL | |         } else {
LL | |             return;
LL | |         }
   | |_________^
```

Reply:
Fixed the output issue. Also fixed rust-lang/rust-clippy#7586. I changed `higher::IfLet` to strictly require that it is not in `while` loop desugaring.

> Another sidenote: The commit is not linked to your GH account (The email probably isn't set). This is not a dealbreaker, just FYI in case that is not intended 🙃

Really appreciate you pointing that out! I have a new machine and didn't have my gitconfig set up right.